lib/python-sdk/common_grants_sdk/utils/transformation.py

```diff
@@ -0,0 +1,128 @@
+"""
+This module provides a utility function for transforming data using a mapping.
+
+The transform_from_mapping function takes a data dictionary and a mapping dictionary.
+The mapping dictionary describes how to transform the data dictionary into a new dictionary.
+"""
+
+from typing import Any, Callable
+
+handle_func = Callable[[dict, dict], Any]
+
+
+def get_from_path(data: dict, path: str, default: Any = None) -> Any:
+    """
+    Gets a value from a dictionary using dot notation.
+
+    Args:
+        data: The dictionary to extract the value from
+        path: A dot-separated string representing the path to the value
+        default: The default value to return if the path doesn't exist
+
+    Returns:
+        The value at the specified path, or the default value if the path doesn't exist
+    """
+    parts = path.split(".")
+    for part in parts:
+        if isinstance(data, dict) and part in data:
+            data = data[part]
+        else:
+            return default
+    return data
+
+
+def handle_field(data: dict, spec: dict) -> Any:
+    """
+    Handles a field transformation by extracting a value from the data using the specified field path.
+
+    Args:
+        data: The source data dictionary
+        spec: A dictionary containing the field specification with a 'field' key
+
+    Returns:
+        The value from the specified field path in the data
+    """
+    return get_from_path(data, spec["field"])
+
+
+def handle_const(_data: dict, spec: dict) -> Any:
+    """
+    Handles a constant transformation by returning the specified constant value.
+
+    Args:
+        data: The source data dictionary (unused)
+        spec: A dictionary containing the constant value under the 'const' key
+
+    Returns:
+        The constant value specified in the spec
+    """
+    return spec["const"]
+
+
+def handle_match(data: dict, spec: dict) -> Any:
+    """
+    Handles a match transformation by looking up a value in a case dictionary.
+
+    Args:
+        data: The source data dictionary
+        spec: A dictionary containing:
+            - 'field': The field path to get the value from
+            - 'case': A dictionary mapping values to their transformations
+            - 'default': (optional) The default value if no match is found
+
+    Returns:
+        The transformed value based on the match, or the default value if no match is found
+    """
+    match_spec = spec["match"]
+    val = get_from_path(data, match_spec["field"])
+    return match_spec["case"].get(val, match_spec.get("default"))
+
+
+# Registry for handlers
+DEFAULT_HANDLERS: dict[str, handle_func] = {
+    "field": handle_field,
+    "const": handle_const,
+    "match": handle_match,
+}
+
+
+def transform_from_mapping(
+    data: dict,
+    mapping: dict,
+    depth: int = 0,
+    max_depth: int = 500,
+    handlers: dict[str, handle_func] = DEFAULT_HANDLERS,
```

Comment:
Allows library users to provide their own custom handlers if they'd like to extend the base set.